src/conn/conn_capacity.c

```diff
@@ -0,0 +1,461 @@
+ * __wt_capacity_server_create --
+ *	Configure and start the capacity server.
+ */
+int
+__wt_capacity_server_create(WT_SESSION_IMPL *session, const char *cfg[])
+{
+	WT_CONNECTION_IMPL *conn;
+
+	conn = S2C(session);
+
+	/*
+	 * If it is a read only connection or if background fsync is not
+	 * supported, then there is nothing to do.
+	 */
+	if (F_ISSET(conn, WT_CONN_READONLY) ||
+	    (__wt_fsync_background_chk(session) == WT_NOTFOUND))
+		return (0);
+
+	/*
+	 * Stop any server that is already running. This means that each time
+	 * reconfigure is called we'll bounce the server even if there are no
+	 * configuration changes. This makes our life easier as the underlying
+	 * configuration routine doesn't have to worry about freeing objects
+	 * in the connection structure (it's guaranteed to always start with a
+	 * blank slate), and we don't have to worry about races where a running
+	 * server is reading configuration information that we're updating, and
+	 * it's not expected that reconfiguration will happen a lot.
+	 */
+	if (conn->capacity_session != NULL)
+		WT_RET(__wt_capacity_server_destroy(session));
+
+	WT_RET(__capacity_config(session, cfg));
+	if (conn->capacity != NULL)
+		WT_RET(__capacity_server_start(conn));
+
+	return (0);
+}
+
+/*
+ * __wt_capacity_server_destroy --
+ *	Destroy the capacity server thread.
+ */
+int
+__wt_capacity_server_destroy(WT_SESSION_IMPL *session)
+{
+	WT_CONNECTION_IMPL *conn;
+	WT_DECL_RET;
+	WT_SESSION *wt_session;
+
+	conn = S2C(session);
+
+	F_CLR(conn, WT_CONN_SERVER_CAPACITY);
+	if (conn->capacity_tid_set) {
+		__wt_cond_signal(session, conn->capacity_cond);
+		WT_TRET(__wt_thread_join(session, &conn->capacity_tid));
+		conn->capacity_tid_set = false;
+	}
+	__wt_cond_destroy(session, &conn->capacity_cond);
+
+	/* Close the server thread's session. */
+	if (conn->capacity_session != NULL) {
+		wt_session = &conn->capacity_session->iface;
+		WT_TRET(wt_session->close(wt_session, NULL));
+	}
+
+	/*
+	 * Ensure capacity settings are cleared - so that reconfigure doesn't
+	 * get confused.
+	 */
+	conn->capacity_session = NULL;
+	conn->capacity_tid_set = false;
+	conn->capacity_cond = NULL;
+	__wt_free(session, conn->capacity);
+
+	return (ret);
+}
+
+/*
+ * __capacity_signal --
+ *	Signal the capacity thread if sufficient data has been written.
+ */
+static void
+__capacity_signal(WT_SESSION_IMPL *session)
+{
+	WT_CAPACITY *cap;
+	WT_CONNECTION_IMPL *conn;
+
+	conn = S2C(session);
+	cap = conn->capacity;
+	if (cap->written >= cap->threshold && !cap->signalled) {
+		__wt_cond_signal(session, conn->capacity_cond);
+		WT_PUBLISH(cap->signalled, true);
+	}
+}
+
+/*
+ * __capacity_reserve --
+ *	Make a reservation for the given number of bytes against
+ * the capacity of the subsystem.
+ */
+static void
+__capacity_reserve(WT_SESSION_IMPL *session, uint64_t *reservation,
+    uint64_t bytes, uint64_t capacity, uint64_t now_ns, uint64_t *result)
+{
+	uint64_t res_len, res_value;
+
+	WT_UNUSED(session);
```

Comment:
Since `__capacity_reserve` is private to this file, why pass in a `WT_SESSION_IMPL` at all?

Reply:
I will let @ddanderson make the call on this one. I agree with this sentiment. But it is there because we originally had debugging/verbose messages in there, and needed the session. Might we need that again sometime?

Reply:
Fixed per @keithbostic's suggestion.